DOCSP-33429: atlas search idx mgmt (#144)
mongodb/docs-kotlin: source/fundamentals/indexes.txt

      name_1_age_-1

   MongoDB uses this index when you sort your data by either:

   - ``name`` ascending, ``age`` descending
   - ``name`` descending, ``age`` ascending

   Specifying a sort order of ``name`` and :guilabel:`age` ascending or :guilabel:`name` and ``age``
   descending requires an in-memory sort.

For more information on how to ensure your index covers your query criteria and projection, see the Server manual
articles on :manual:`query coverage </core/query-optimization/#covered-query>`.

Operational Considerations
~~~~~~~~~~~~~~~~~~~~~~~~~~

> To improve query performance, build indexes on fields that appear often in
> your application's queries and operations that return sorted results. Each
> index that you add consumes disk space and memory when active, so we recommend
> that you track index memory and disk usage for capacity planning. In addition,
> when a write operation updates an indexed field, MongoDB updates the related
> index.

Reviewer: S: i have trouble connecting the reasons and recommendations here. what about something like this:

```suggestion

You can optimize your application's use of indexes by following these guidelines:

- Track index memory and disk usage for capacity planning. Each index that you add consumes disk space and memory when active.
- To improve query performance, build indexes on fields that appear often in
your application's queries and operations that return sorted results. 
- Avoid adding indexes to rarely used fields. When a write operation updates an indexed field, MongoDB also updates the related index.
```

Edit as you like!

Operational Considerations
~~~~~~~~~~~~~~~~~~~~~~~~~~

> The following guidelines describe how you can optimize the way
> application uses indexes:

Reviewer: ```suggestion
The following guidelines describe how you can optimize the way
your application uses indexes:
```

Index Types
-----------

MongoDB supports several different index types to support querying your data. The following sections describe the
most common index types and provide sample code for creating each index type. For a full list of index types, see
:manual:`Indexes </indexes/>` in the Server manual.

.. tip::

>    The {+driver-short+} provides the `Indexes <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/Indexes.html>`__ class that

Reviewer: Q: Are there multiple Index classes, and the driver provides this one? Or is there just one, and this sentence describes what it does?

If the latter, it should be "class, which" instead of "that"

Index Types
-----------

MongoDB supports several different index types to support querying your data. The following sections describe the
most common index types and provide sample code for creating each index type. For a full list of index types, see
:manual:`Indexes </indexes/>` in the Server manual.

.. tip::

   The {+driver-short+} provides the `Indexes
   <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/Indexes.html>`__
   class to create and manage indexes. This class includes static
   factory methods to create index specification documents for different

>    MongoDB Index key types.

Reviewer: Q: lower case?

```suggestion
   MongoDB index key types.
```

      Index created: rated_1_genres_1_title_1

The following is an example of a query that is covered by the index
created in the preceding code snippet:

.. literalinclude:: /examples/generated/IndexesTest.snippet.multikey-index-query.kt
   :language: kotlin

> Multikey indexes behave differently from other indexes in terms of query coverage, index bound computation, and

Reviewer: ```suggestion
Multikey indexes behave differently from other indexes in terms of query coverage, index-bound computation, and
```

sort behavior. To learn more about multikey indexes, including a discussion of their behavior and limitations,

> see the :manual:`Multikey Indexes page </core/index-multikey>` in the MongoDB manual.

Reviewer: S: "Server manual" for consistency

```suggestion
see the :manual:`Multikey Indexes page </core/index-multikey>` in the Server manual.
```

.. _kotlin-search-indexes:

Atlas Search Indexes
~~~~~~~~~~~~~~~~~~~~

The Atlas Search feature enables you to perform full-text searches on
collections hosted on MongoDB Atlas. The indexes specify the behavior of

> the search and which fields to index.

Reviewer: I: This sentence doesn't make sense to me--how indexes specify which fields to index. Can you clarify?

To learn more about MongoDB Atlas Search, see the
:atlas:`Atlas Search Indexes </atlas-search/atlas-search-overview/#fts-indexes>`
documentation.

You can call the following methods on a collection to manage your Atlas Search
indexes:

- ``createSearchIndex()``
- ``createSearchIndexes()``
- ``listSearchIndexes()``
- ``updateSearchIndex()``
- ``dropSearchIndex()``

.. note::

>    The Atlas Search Index management methods run asynchronously. The

Reviewer: S: Guessing "Atlas Search Index" isn't a proper noun phrase? Would lowercase + hyphenate the compound modifier

```suggestion
   The Atlas Search index-management methods run asynchronously. The
```

   driver methods can return before confirming that they ran
   successfully. To determine the current status of the indexes, call the
   ``listSearchIndexes()`` method.

The following sections provide code examples that demonstrate how to use
each of the preceding methods.

Create a Search Index
+++++++++++++++++++++

> You can use the `createSearchIndex() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/create-search-index.html>`__
> and the
> `createSearchIndexes() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/create-search-indexes.html>`__

Reviewer: S: ditch the first "the" if you're using "method" just once

```suggestion
You can use the `createSearchIndex() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/create-search-index.html>`__
and
`createSearchIndexes() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/create-search-indexes.html>`__
```

methods to create Atlas Search indexes.

The following code example shows how to create a single index:

.. literalinclude:: /examples/generated/SearchIndexesTest.snippet.single-search-index-create.kt
   :language: kotlin

The following code example shows how to create multiple indexes:

.. literalinclude:: /examples/generated/SearchIndexesTest.snippet.multi-search-index-create.kt
   :language: kotlin

List Search Indexes
+++++++++++++++++++

You can use the
`listSearchIndexes() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/list-search-indexes.html>`__
method to return the Atlas Search indexes of a collection.

The following code example shows how to print a list of the search indexes of

> a collection:

Reviewer: S: "indexes of a collection" vs. "indexes for a collection"? 

mentioning here, but i trust your judgment!

Author: on

and `createSearchIndexes() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/create-search-indexes.html>`__
methods to create Atlas Search indexes on a collection.

The following code example shows how to create a single index:

.. literalinclude:: /examples/generated/SearchIndexesTest.snippet.single-search-index-create.kt
   :language: kotlin

The following code example shows how to create multiple indexes:

.. literalinclude:: /examples/generated/SearchIndexesTest.snippet.multi-search-index-create.kt
   :language: kotlin

List Search Indexes
+++++++++++++++++++

You can use the
`listSearchIndexes() <{+api+}/apidocs/mongodb-driver-kotlin-coroutine/mongodb-driver-kotlin-coroutine/com.mongodb.kotlin.client.coroutine/-mongo-collection/list-search-indexes.html>`__
method to return a list of the Atlas Search indexes on a collection.

The following code example shows how to print a list of the search indexes of

> a collection:

Reviewer: Based on [your comment](https://github.com/mongodb/docs-kotlin/pull/144#discussion_r1453607066), I thought you were changing from "of" to "on" (just making sure you didn't intend to make this change)

Author: AH, changed in one place but not the other.

MongoDB supports queries of geospatial coordinate data using **2dsphere indexes**. With a ``2dsphere`` index, you can query
the geospatial data for inclusion, intersection, and proximity. For more information on querying geospatial data, see
:manual:`Geospatial Queries </geospatial-queries/>` in the Server manual.

To create a ``2dsphere`` index, you must specify a field that contains
only **GeoJSON objects**. To learn more about this

> type, see the :manual:`GeoJSON objects reference </reference/geojson>`

Reviewer: ```suggestion
type, see :manual:`GeoJSON objects</reference/geojson>`
```

      Index created: location.geo_2dsphere

.. important::

>    Attempting to create a geospatial index on a field that is covered by a geospatial index results in an error.

Reviewer: S: Clearer to include "already"? I assume that's what this means

```suggestion
   Attempting to create a geospatial index on a field that is already covered by a geospatial index results in an error.
```
